src/stopping_criterion.jl

```diff
@@ -0,0 +1,615 @@
+    is_finished!(problem::Problem, algorithm::Algorithm, state::State)
+    is_finished!(problem::Problem, algorithm::Algorithm, state::State, stopping_criterion::StoppingCriterion, stopping_criterion_state::StoppingCriterionState)
+
+Indicate whether an [`Algorithm`](@ref) solving [`Problem`](@ref) is finished having reached
+a certain [`State`](@ref). The variant with three arguments by default extracts the
+[`StoppingCriterion`](@ref) and its [`StoppingCriterionState`](@ref) and their actual
+checks are performed in the implementation with five arguments.
+
+The mutating variant does alter the `stopping_criterion_state` and and should only be called
+once per iteration, the other one merely inspects the current status without mutation.
+"""
+
+@doc "$(_doc_is_finished)"
+function is_finished(problem::Problem, algorithm::Algorithm, state::State)
+    return is_finished(
+        problem,
+        algorithm,
+        state,
+        algorithm.stopping_criterion,
+        state.stopping_criterion_state,
+    )
+end
+
+@doc "$(_doc_is_finished)"
+is_finished(::Problem, ::Algorithm, ::State, ::StoppingCriterion, ::StoppingCriterionState)
+
+@doc "$(_doc_is_finished)"
+function is_finished!(problem::Problem, algorithm::Algorithm, state::State)
+    return is_finished!(
+        problem,
+        algorithm,
+        state,
+        algorithm.stopping_criterion,
+        state.stopping_criterion_state,
+    )
+end
+
+@doc "$(_doc_is_finished)"
+is_finished!(::Problem, ::Algorithm, ::State, ::StoppingCriterion, ::StoppingCriterionState)
+
+@doc """
+    summary(io::IO, stopping_criterion::StoppingCriterion, stopping_criterion_state::StoppingCriterionState)
+
+Provide a summary of the status of a stopping criterion – its parameters and whether
+it currently indicates to stop. It should not be longer than one line
+
+# Example
+
+For the [`StopAfterIteration`](@ref) criterion, the summary looks like
+
+```
+Max Iterations (15): not reached
+```
+"""
+Base.summary(io::IO, ::StoppingCriterion, ::StoppingCriterionState)
+
+#
+#
+# Meta StoppingCriteria
+@doc raw"""
+    StopWhenAll <: StoppingCriterion
+
+store a tuple of [`StoppingCriterion`](@ref)s and indicate to stop,
+when _all_ indicate to stop.
+
+# Constructor
+
+    StopWhenAll(c::NTuple{N,StoppingCriterion} where N)
+    StopWhenAll(c::StoppingCriterion,...)
+"""
+struct StopWhenAll{TCriteria<:Tuple} <: StoppingCriterion
+    criteria::TCriteria
+end
+StopWhenAll(c::AbstractVector{<:StoppingCriterion}) = StopWhenAll(Tuple(c))
+StopWhenAll(c...) = StopWhenAll(c)
+function indicates_convergence(stop_when_all::StopWhenAll)
+    return any(indicates_convergence, stop_when_all.criteria)
+end
+
+function Base.show(io::IO, ::MIME"text/plain", stop_when_all::StopWhenAll)
+    print(io, "StopWhenAll with the Stopping Criteria")
+    for stopping_criterion in stop_when_all.criteria
+        println(io)
+        replace(io, string(stopping_criterion), "\n" => "\n    ") #increase indent
```

Comment:
I am not so sure, but for me this reads like:

For the first SC in the group it adds 4 spaces after a new line
For the second _and_ the first (all of the content of the IO buffer?) it changed `\n` to add 4 spaces.

Might that then look like

```
            A
        B
    C
normal text
```
(when A, B, C are the prints of the children stopping criteria)?
I only meant to indent A, B, and C by 4 spaces :)

Reply:
Ah, I thought replace would go through all the buffer. Then it should be fine. But you would have noticed in a test anyways ;)